File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) {
        if(head == NULL) return head;
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        unordered_map<int, ListNode*>mp;
        int sum = 0;
        mp[sum] = dummy;
        ListNode* cur = head;
        while(cur) {
            sum += cur->val;
            if(mp.find(sum) != mp.end()) {
                ListNode* prev = mp[sum]->next;
                int temp_sum = sum + prev->val;
                while(temp_sum != sum) {
                    mp.erase(temp_sum);
                    prev = prev->next;
                    temp_sum += prev->val;
                }
                mp[sum]->next = cur->next;
            }
            else {
                mp[sum] = cur;
            }
            cur = cur->next;
        }
        return dummy->next;
    }
};

File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        if(head == null) return null;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        Map<Integer, ListNode>mp = new HashMap<>();
        int sum = 0;
        mp.put(0, dummy);
        ListNode cur = head;
        while(cur != null) {
            sum += cur.val;
            if(mp.containsKey(sum)) {
                ListNode prev = mp.get(sum).next;
                System.out.println(sum);
                int temp_sum = sum + prev.val;
                while(temp_sum != sum) {
                    mp.remove(temp_sum);
                    prev = prev.next;
                    temp_sum += prev.val;
                }
                mp.get(sum).next = prev.next;
            }
            else {
                mp.put(sum, cur);
            }
            cur = cur.next;
        }
        return dummy.next;
    }
}